cmd/cloud_sql_proxy/proxy.go

```diff
@@ -273,6 +274,15 @@ func parseInstanceConfig(dir, instance string, cl *http.Client) (instanceConfig,
 	if err != nil {
 		return instanceConfig{}, err
 	}
+
+	// Check to see if one of the addresses is public or not
+	// and mark that on the config
+	for _, mapping := range inst.IpAddresses {
+		if mapping.Type == "PRIMARY" {
+			ret.IsPublic = true
```

Comment:
What if they've specified private as a preferred IP type? Will this still return public?

Reply:
Not sure I follow. For private instances, this should be false since that's the default value in `Go` from what I can tell.

Reply:
A Cloud SQL instance can have both a public and private ip. By default, the proxy uses a PUBLIC ip if it exists, otherwise a PRIVATE one. Users can override this behavior by specifying a different preference using [`-ip_address_types`]. If I'm reading this correctly, the current logic doesn't take this into account, meaning it will mark an instance as public even if the private ip is used to dial.

Reply:
I get what you mean. I'm going to add another flag and try and reconcile it with a nicer error message. 

Not sure what else; we could try and provide both? But the user may not care about checking both perse.

Reply:
Done. Two queries:

1. Is this OK or should we do something else with this?
2. Is there an easy way for me to write some unit tests for this?

Reply:
I don't think this approach is probably the right way to approach this either. I think probably some better approaches would be one of the following:

1. Reimplement similar logic to [`findIpAddr`](https://github.com/GoogleCloudPlatform/cloudsql-proxy/blob/3acc260458d1198c32060d20eadfa36cce723a23/proxy/certs/certs.go#L208) here in cmd, except returns the IP type used rather than the addr  (`findIpType`?). 

2. Modify `findIpAddr` to return addr and type, and then move the "test connectivity" function into the client so that it has access to it.

Option 2 is probably better since we don't have to reimplement logic and don't run the risk that one changes but not the other, but might be somewhat invasive. I'm not sure without taking a closer look.

Reply:
There is no obvious reason that `findIpAddr` has to be in `certs` either from what I can tell since it is free most of dependencies inside of that module. You could probably split it out into something else.

That being said, at what cost? Does the user really need to know public / private? They're going to get the IP in the actual exception anyway so they could go check. Do you think there is enough value add here? 

The hairy part is probably the "cert source" structures and splitting those out.

Thoughts?

Reply:
At some point, we definitely want to do this. One of the bigger issues with the proxy is that folks don't understand it's not a VPN, and doesn't grant a connection to an instance via Private IP if it didn't have it before. As some point I'd like to be able to provide troubleshooting steps specific to the IP type to make the error actionable (Using private IP and can't connect - Is the firewall egress rules allow for port 3307? Are you on a VPC?) and maybe even add some diagnostic info (like checking for a VPC metadata server to confirm connectivity).

That said, having this info included as part of this PR is "nice to have" but not required - we can always expand upon it later.